library/sm3.c

```diff
@@ -0,0 +1,429 @@
+/*
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+/*
+ *  GB/T 32905-2016 compliant SM3 implementation
+ *  The SM3 algorithm was designed by Xiaoyun Wang et al in 2010.
+ *
+ *  http://www.gmbz.org.cn/upload/2018-07-24/1532401392982079739.pdf
+ */
+
+#include "common.h"
+
+#if defined(MBEDTLS_SM3_C)
+
+#include "mbedtls/sm3.h"
+#include "sm3_internal.h"
+#include "mbedtls/platform_util.h"
+#include "mbedtls/error.h"
+
+#include <string.h>
+
+#if defined(MBEDTLS_SELF_TEST)
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdio.h>
+#define mbedtls_printf printf
+#endif /* MBEDTLS_PLATFORM_C */
```

Comment:
We've realized that the conditional inclusion of `platform.h` was not a good idea, and [we're moving to unconditional inclusion](https://github.com/Mbed-TLS/mbedtls/pull/6291). Please use the new style in this new file.
```suggestion
#include "mbedtls/platform.h"
```
Or even include `mbedtls/platform.h` regardless of `MBEDTLS_SELF_TEST`.